sklearn/linear_model/sag_fast.pyx

```diff
@@ -32,49 +27,219 @@ cdef inline double fmax(double x, double y) nogil:
     return y
 
 
+cdef double _logsumexp(double* arr, int n_classes) nogil:
+    """Computes the sum of arr assuming arr is in the log domain.
+
+    Returns log(sum(exp(arr))) while minimizing the possibility of
+    over/underflow.
+    """
+    # Use the max to normalize, as with the log this is what accumulates
+    # the less errors
+    cdef double vmax = arr[0]
+    cdef double out = 0.0
+    cdef int i
+
+    for i in range(1, n_classes):
+        if vmax < arr[i]:
+            vmax = arr[i]
+
+    for i in range(n_classes):
+        out += exp(arr[i] - vmax)
+
+    return log(out) + vmax
+
+
+cdef class MultinomialLogLoss:
```

Comment:
Is it worthy to mention "The gradient of the multinomial logistic loss for every sample" is, (thought that is kind of obvious here)